File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque
read = sys.stdin.readline
n, e, v = map(int, read().split())
edge = [[] for _ in range(n+1)]
visited = [False] * (n+1)
for _ in range(e):
    a, b = map(int, read().split())
    edge[a].append(b)
    edge[b].append(a)
for i in range(len(edge)):
    edge[i].sort()
    
   
def dfs(v, visited, arr):
    if visited[v]:
        return
    visited[v] = True
    print(v, end=' ')
    for next_v in arr[v]:
        if not visited[next_v]:
            dfs(next_v, visited, arr)
    return


def bfs(v, arr):
    queue = deque()
    queue.append(v)
    visited = [False] * (n+1)
    visited[v] = True
    while queue:
        v = queue.popleft()
        print(v, end=' ')
        for next_v in arr[v]:
            if not visited[next_v]:
                visited[next_v] = True
                queue.append(next_v)
    return

               
dfs(v, visited, edge)
print()
bfs(v, edge)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.util.*;
import java.io.*;

class Main {
    static int n, m, v;
    static StringBuilder sb = new StringBuilder();
    static ArrayList<Integer>[] edges;
    static boolean[] visited;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        v = Integer.parseInt(st.nextToken()) - 1;
        edges = new ArrayList[n];
        for (int i = 0; i < n; i++) edges[i] = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken()) - 1;
            int b = Integer.parseInt(st.nextToken()) - 1;
            edges[a].add(b);
            edges[b].add(a);
        }
        for (int i = 0; i < n; i++) edges[i].sort(null);
        
        visited = new boolean[n];
        visited[v] = true;
        dfs(v);
        sb.append("\n");
        bfs(v);
        System.out.println(sb);
    }
    
    static void dfs(int v) {
        sb.append(v+1).append(" ");
        for (int nv : edges[v]) {
            if (!visited[nv]) {
                visited[nv] = true;
                dfs(nv);
            }
        }
    }
    
    static void bfs(int v) {
        visited = new boolean[n];
        visited[v] = true;
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        queue.add(v);
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            sb.append(cur + 1).append(" ");
            for (int nv : edges[cur]) {
                if (!visited[nv]) {
                    visited[nv] = true;
                    queue.add(nv);
                }
            }
        }
    }
}